### DISCUSSION TITLE
DiffusionFluxAux & Functor Material Properties

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#20235] [Q&A General] [answered]
Hi:

I am using the NS module with the FV formulation and WCNSFV to run cases with natural convection in a cavity (code converges great!).  For the fluid properties, I am using the Nitrogen fp and making them available using GeneralFunctorFluidProps.

I want to have a VPP for the distribution of the heat flux on the sides of the cavity.  So, I tried using DiffusionFluxAux.  However, this reports the diffusivity 'k' as not existing.  I assume this is because NS::k is a functor and not a traditional material property.  Is that correct?  Do we need new versions of objects such as DiffusionFluxAux that can accept functors?

I tried generating a parallel set of nitrogen properties as normal materials using GeneralFluidProps (adding a suffix to distinguish them).  This didn't work because one of the properties wasn't defined (e_from_something).  Is there anyway to get just the one property that I need?

Note, this is for a user that is simply running a Moose-based app, not someone that wants to modify and rebuild the code.

Thanks,
  Joe

**Comment by GiudGiud** [comment, marked as answer]
Hi Joe

you are right that the issue is with k not being a functor. 
i had to do something similar for @muelcm and the way we got to the diffusion flux was through the GenericFunctorGradientAux (or something similar as a functor matprop), with the diffusivity just being the ‘factor’. 

I m out today but I ll post a snippet when I come back. 

Guillaume

**Reply by GiudGiud** [reply]
Ok so the workflow I came up with:

- use first a GenericFunctorGradientMaterial to get the gradient of the temperature field in a vector function property
- use then a ADFunctorElementalVectorAux on each component with the gradient of the temperature as the functor and the diffusivity as the factor. This will populate the auxvariables with each component of the diffusive flux. This will not handle anisotropic diffusion coefficients. Please let me know if you need that. 
- use the vectorMagnitudeAux to compute the magnitude of the flux.

This gives an "elemental" flux, computed at the centroid of the elements, similarly to the DiffusionFluxAux. It's not exactly the 'true' finite volume flux, which is really computed on the faces of the elements. As we introduce more functionality for face variables into MOOSE, we'll be able to output that, in a face variable. There should be some progress on this by the end of this FY. 

Please let me know if that does not work for you
